include/matrix_assembler.cc

```diff
@@ -8,7 +8,7 @@ MatrixAssembler::MatrixAssembler() {
 
 // Obtain global nodal and pipe indices and pointers from meshes
 void MatrixAssembler::global_nodal_pipe_indices(
-    const std::shared_ptr<Mesh>& mesh) {
+    const std::unique_ptr<Mesh>& mesh) {
```

Comment:
You are changing the ownership of mesh, we shouldn't be doing this. This will require a std::move to accomplish.  Have a look at https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/